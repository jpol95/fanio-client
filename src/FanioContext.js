import React from 'react'

const FanioContext = React.createContext({
    fandomList: [],
    reviewList: [],
    sectionList: [], 
    subList: [],
    tagList: [],
    reviewTagList: [],
    installmentList: [],
    handleSubmitReview: () => {}, 
    handleAddFandom: () => {}, 
    handleSubmitInstallments: () => {}, 
    handleGetReview: () => {}
})

export default FanioContext