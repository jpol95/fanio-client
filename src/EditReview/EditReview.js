import React from "react";
import FanioContext from "../FanioContext";
import FetchService from "../FetchService";
import ReviewForm from '../ReviewForm/ReviewForm'

export default class EditReview extends React.Component {
  static contextType = FanioContext;
  tableName = this.props.match.params.subId ? "sub" : "section";
  sectionId = this.props.match.params.sectionId;
  subId = this.props.match.params.subId;
  fandomId = Number(this.props.match.params.fandomId);
  installmentId = this.props.match.params.installmentId;
  userId = Number(this.props.match.params.userId);
  reviewId = Number(this.props.match.params.reviewId);


  handleSubmit = async (e, reviewObject) => {
    e.preventDefault();
    const link =
      this.tableName === "subs"
        ? `/users/${this.userId}/fandoms/${this.fandomId}/installments/${this.installmentId}/sections/${this.sectionId}/subs/${this.subId}/review/${this.reviewId}`
        : `/users/${this.userId}/fandoms/${this.fandomId}/installments/${this.installmentId}/sections/${this.sectionId}/review/${this.reviewId}`;
    const {title, content, rating} = reviewObject
    const reviewToPost = {title, content, rating, id: this.reviewId }
    const review = await FetchService.patchReview(reviewToPost);
    await FetchService.deleteTrelsByReview(review.id)
    const trelsToPost = reviewObject.tags.map(tag => {return {tagId: tag, reviewId: review.id}} )
    const newTrels = await FetchService.postTrels(trelsToPost)
    this.props.history.push(link);
    this.context.handleEditReview(review, newTrels);
  };

  getReviewById = () => {
    return this.context.reviewList.find((review) => review.id === this.reviewId);
  };

  getTrels = () => {
      return this.context.reviewTagList.filter(trel => trel.reviewId === this.reviewId)
      .map(trel => trel.tagId)
  }

  render() {
      const review = {...this.getReviewById(), tags: this.getTrels()}
    return (
      <>
        <h3>Edit review</h3>
        <ReviewForm {...review} handleSubmit={this.handleSubmit} />
      </>
    );
  }
}
