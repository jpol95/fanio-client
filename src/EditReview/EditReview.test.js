import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import EditReview from './EditReview';
import FanioContext from '../FanioContext'
import dummyStore from '../dummy-store'

describe('Edit Review component', () => {
    const props = {
    match: {
        params: {
            userId: 1, 
            fandomId: 1, 
            installmentId: 1, 
            sectionId: 1, 
            subId: 1, 
            reviewId: 13
        }
    }, 
    installId: 1, 
}


  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<FanioContext.Provider value={{reviewList : dummyStore.reviewList, reviewTagList: dummyStore.reviewTagList, tagList: dummyStore.tagList}}> <EditReview {...props} /> </FanioContext.Provider>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the UI as expected', () => {
    const tree = renderer
      .create(<FanioContext.Provider value={{reviewList : dummyStore.reviewList, reviewTagList: dummyStore.reviewTagList, tagList: dummyStore.tagList}}> <EditReview {...props} /> </FanioContext.Provider>)
      .toJSON();
    expect(tree).toMatchSnapshot();  
  });
});