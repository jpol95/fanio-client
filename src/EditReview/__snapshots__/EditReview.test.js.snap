// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Review component renders the UI as expected 1`] = `
Array [
  " ",
  <h3>
    Edit review
  </h3>,
  <form
    className="create"
    onSubmit={[Function]}
  >
    <label
      htmlFor="title"
    >
      Title
    </label>
    <input
      defaultValue="Supernatural season 2 Episode 7 Review"
      id="title"
      onChange={[Function]}
      type="text"
    />
    <label
      htmlFor="rating"
    >
      Rating
    </label>
    <input
      defaultValue={4}
      id="rating"
      onChange={[Function]}
      type="text"
    />
    <label
      htmlFor="content"
    >
      Review
    </label>
    <textarea
      defaultValue="Omg this is the worst episode ever!"
      id="content"
      onChange={[Function]}
      type="text"
    />
    <label
      htmlFor="tag-input"
    >
      Tags
    </label>
    <input
      id="tag-input"
      list="tags"
      onChange={[Function]}
    />
    <button
      className="add-tag-button"
      onClick={[Function]}
    >
      Add Tag
    </button>
    <datalist
      id="tags"
      type="text"
    >
      <option
        id={1}
        value="science-fiction"
      />
      <option
        id={2}
        value="comedy"
      />
      <option
        id={3}
        value="drama"
      />
      <option
        id={4}
        value="period"
      />
      <option
        id={5}
        value="romance"
      />
      <option
        id={6}
        value="historical-fiction"
      />
      <option
        id={7}
        value="fantasy"
      />
      <option
        id={8}
        value="raunchy"
      />
      <option
        id={9}
        value="family"
      />
      <option
        id={10}
        value="corny"
      />
      <option
        id={11}
        value="violent"
      />
      <option
        id={12}
        value="horror"
      />
    </datalist>
    <br />
    <div
      className="tag-list"
    />
    <button
      className="submit-review-button"
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </form>,
  " ",
]
`;
